### ISSUE TITLE
Error in Pipeline: troublet call

### ISSUE THREAD

**Issue by neilrobertson** [wheaton5/souporcell#38]
Hi! Excuse my potential for stupidity, but I'm currently running into the same problem across a couple of datasets at the doublet detection stage below:

```
singularity exec souporcell.sif souporcell_pipeline.py -i ./CONTROL_SS01.possorted_genome_bam.bam -b ./CONTROL_SS01.barcodes.tsv -f ./cellranger_GRCh38.genome.fa -t 40 -k 3 -o ./CONTROL_SS01/ --ploidy 2 --min_alt 10 --min_ref 10 --common_variants ./filtered_2p_1kgenomes_GRCh38.vcf
checking modules
imports done
checking bam for expected tags
checking fasta
restarting pipeline in existing directory ./CONTROL_SS01/
no bam index found, creating
generating fastqs with cell barcodes and umis in readname
remapping with minimap2

cleaning up tmp fastqs
repopulating cell barcode and UMI tags
sorting retagged bam files
merging sorted bams
cleaning up tmp samfiles
using common variants
40
[main_samview] region "1:77493769-77493768" could not be parsed. Continue anyway.
[main_samview] region "1:154987538-77493768" could not be parsed. Continue anyway.
[main_samview] region "2:107786145-77493768" could not be parsed. Continue anyway.
[main_samview] region "3:91088843-77493768" could not be parsed. Continue anyway.
[main_samview] region "5:90047574-77493768" could not be parsed. Continue anyway.
running vartrix
running souporcell clustering
/opt/souporcell/souporcell/target/release/souporcell -k 3 -a ./CONTROL_SS01//alt.mtx -r ./CONTROL_SS01//ref.mtx --restarts 100 -b ./CONTROL_SS01.barcodes.tsv --min_ref 10 --min_alt 10 --threads 40
running souporcell doublet detection
Traceback (most recent call last):
  File "/opt/souporcell/souporcell_pipeline.py", line 535, in <module>
    doublets(args, ref_mtx, alt_mtx, cluster_file)
  File "/opt/souporcell/souporcell_pipeline.py", line 481, in doublets
    subprocess.check_call(["troublet", "--alts", alt_mtx, "--refs", ref_mtx, "--clusters", cluster_file], stdout = dub, stderr = err)
  File "/usr/local/envs/py36/lib/python3.6/subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['troublet', '--alts', './CONTROL_SS01//alt.mtx', '--refs', './CONTROL_SS01//ref.mtx', '--clusters', './CONTROL_SS01//clusters_tmp.tsv']' returned non-zero exit status 101.
```

I was wondering if you could help me troubleshoot - I've ran on a couple of different machines/datasets but keep bumping into this...

Thanks again!

**Comment by wheaton5** [comment]
There should be a doublets.err file in your output directory. Can you give me the output from that?

Also can you tell me a little about these samples? That might give me more insight into potential problems.

Best,
Haynes

**Comment by neilrobertson** [comment]
Hi Haynes,

Many thanks for helping me!  This is a rather simple experiment with three pooled cell-lines ran in 10x.

I’ve copied the required files to the directory as so, and ran pipeline:

cd /mnt/tchandra-lab/Neil/projects/Gab_scFibroblast_Genotyping/Souporcell
mkdir ./CONTROL_SS01/
cp /mnt/tchandra-lab/Gab/data/scRNA-seq/Chandra_Etoposide_3CellLines/count/CONTROL_SS01/outs/possorted_genome_bam.bam ./CONTROL_SS01.possorted_genome_bam.bam
cp /mnt/tchandra-lab/Gab/data/scRNA-seq/Chandra_Etoposide_3CellLines/count/CONTROL_SS01/outs/filtered_feature_bc_matrix/barcodes.tsv ./CONTROL_SS01.barcodes.tsv
cp /mnt/tchandra-lab/Gab/genomes/GRCh38/refdata-cellranger-GRCh38-3.0.0/fasta/genome.fa ./cellranger_GRCh38.genome.fa
cp /mnt/tchandra-lab/Gab/genomes/GRCh38/refdata-cellranger-GRCh38-3.0.0/fasta/genome.fa.fai ./cellranger_GRCh38.genome.fa.fai


cd /mnt/tchandra-lab/Neil/projects/Gab_scFibroblast_Genotyping/Souporcell
singularity exec souporcell.sif souporcell_pipeline.py \
-i ./CONTROL_SS01.possorted_genome_bam.bam \
-b ./CONTROL_SS01.barcodes.tsv \
-f ./cellranger_GRCh38.genome.fa \
-t 40 \
-k 3 \
-o ./CONTROL_SS01/ \
--ploidy 2 \
--min_alt 10 \
--min_ref 10 \
--common_variants ./filtered_2p_1kgenomes_GRCh38.vcf

The output from doublets.err is here:

6082 loaded 0 counts, is this a problem?
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /rustc/f3e1a954d2ead4e2fc197c7da7d71e6c61bad196/src/libcore/macros/mod.rs:15<http://mod.rs:15>:40
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

I hope this is clear enough!

Thanks again,

Neil

On 19 Mar 2020, at 06:42, Haynes Heaton <notifications@github.com<mailto:notifications@github.com>> wrote:



There should be a doublets.err file in your output directory. Can you give me the output from that?

Also can you tell me a little about these samples? That might give me more insight into potential problems.

Best,
Haynes

**Comment by drneavin** [comment]
I am also receiving this error in my runs and it seems to be related to vartrix not running correctly through the pipeline. My alt.mtx and ref.mtx matrices were empty except the header lines when I run souporcell_pipeline.py but there was no error file to indicate anything had gone wrong. When I run vartrix with 
```
singularity exec $BUCKETS/AllSoftwares.sif vartrix --umi --mapq 30 -b $BAM -c $BARCODES --scoring-method coverage --threads 8 --ref-matrix $OUT/$stage/ref.mtx --out-matrix $OUT/$stage/alt.mtx -v $OUT/$stage/souporcell_merged_sorted_vcf.vcf.gz --fasta $FASTA
```
vartrix works correctly and I receive the correct alt.mtx and ref.mtx. Then I can run souporcell with
```
singularity exec $BUCKETS/AllSoftwares.sif souporcell_pipeline.py -i $BAM -b $BARCODES -f $FASTA -o $OUT/$stage -k $N -t 8
```
which picks up the pipeline after vartrix and correctly identifies the cells. As a note, I had run the same sample with a previously downloaded version of the singularity sif. The results were the same between the old and new runs but with the old image I didn't receive this error and it completed correctly - seems to be an issue with a newer production of the singularity image.

**Comment by wheaton5** [comment]
This seems very likely related to another issue I have fixed, but am waiting for a few tests to come back before uploading. I should be able to upload a new singularity file tomorrow if those tests come back successful. Thank you and I'm very sorry for this problem.

**Comment by drneavin** [comment]
Thanks @wheaton5! I really appreciate your continued support on souporcell.

**Comment by neilrobertson** [comment]
Thanks @wheaton5! Looking forward to it!

**Comment by wheaton5** [comment]
fix live

**Comment by neilrobertson** [comment]
Many thanks @wheaton5, it's working well and results with common_genotypes are looking great!

If you don't mind, can I ask about the --known_genotypes functionality: does the order or the clusters reflect the order of the input vcf?  Perhaps I'm using this incorrectly, however, the --known_genotypes_sample_names argument seems somewhat redundant as I don't see it reflected in the output?

```
singularity exec souporcell.sif souporcell_pipeline.py \
-i ./CONTROL_SS01/CONTROL_SS01.possorted_genome_bam.bam \
-b ./CONTROL_SS01/CONTROL_SS01.barcodes.tsv \
-f ./cellranger_GRCh38.genome.fa \
-t 20 \
-k 3 \
-o ./CONTROL_SS01/fibroblast_vars_k3/ \
--ploidy 2 \
--min_alt 10 \
--min_ref 10 \
--known_genotypes ./fibroblast_lines.freebayes.pop.no_chr.vcf \
--known_genotypes_sample_names BJ IMR90 WI_38
```
Thanks again, it really is a great tool!

Neil

**Comment by wheaton5** [comment]
Sorry. I meant to make an update to the output for known genotypes, but it hit some complications and i was already behind promises I had made to users.

The order of the clusters should be the order of the --known_genotypes_sample_names even though they are numbered in the output instead of named.

Sorry for this confusion and I will fix this in the near future. Please let me know if you have more questions.

Best,
Haynes

**Comment by neilrobertson** [comment]
Many thanks @wheaton5!

**Comment by SharonWang** [comment]
@wheaton5 
I downloaded the singularity from the link below:
wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --quiet --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=124sRtZoDlKt-jJYS6BbWGC1-CY01aAGT' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=124sRtZoDlKt-jJYS6BbWGC1-CY01aAGT" -O souporcell.sif && rm -rf /tmp/cookies.txt

And I have come across the same error msgs that doublet detection failed. I am wondering if the singularity has been updated already so that the bug has been fixed. If so would it be possible to know where I can get it downloaded.

In addition, although the clustering ran without giving me errors but both of the output files clusters.tsv and clusters_tmp.tsv are empty. Would it be possible to know why this happens?

Best regards,

Xiaonan Wang

**Comment by wheaton5** [comment]
Are you restarting it in the same directory or starting a new run with a new output directory?

**Comment by wheaton5** [comment]
That is the correct singularity download.

**Comment by SharonWang** [comment]
> Are you restarting it in the same directory or starting a new run with a new output directory? Just checking because if pointed at a partially complete output directory, it will use the work done already as a baseline and it sounds like it was the faulty output of previous steps which is making the troublet call fail.

I ran it in a new output directory.  I can see that the following files in the output directory:

    0 Apr  2 14:38 clustering.done
    0 Apr  2 14:38 clusters.tsv
  246 Apr  2 14:38 doublets.err
 6.5K Apr  2 14:38 clusters.err
    0 Apr  2 14:38 clusters_tmp.tsv
    0 Apr  2 14:38 vartrix.done
  67M Apr  2 14:38 ref.mtx
  67M Apr  2 14:38 alt.mtx
   54 Apr  2 14:34 variants.done
 107K Apr  2 14:34 souporcell_merge
   77 Apr  2 14:34 bcftools.err
  17M Apr  2 14:34 souporcell_merge
    0 Apr  2 14:30 retagging.done
 4.4M Apr  2 14:29 souporcell_minim
 8.9G Apr  2 14:26 souporcell_minim
 1.5K Apr  2 13:52 retag.err
 1.2K Apr  2 13:46 remapping.done
 1.7K Apr  2 13:46 minimap.err
 9.4K Apr  2 12:44 fastqs.done

It seems that the variants calling worked as ref.mtx and alt.mtx are not empty but somehow the next clustering step failed. But it did not give any error msgs. It will be very appreciated if you could please give me some directions why step I should rerun?

Best regards,
Xiaonan Wang

**Comment by drneavin** [comment]
@wheaton5 and @SharonWang , I can confirm that on my end this issue was resolved after I redownloaded the sif file yesterday. I was receiving that error until the most recent release. @SharonWang, have you redownloaded the sif file in the last 48 hours?

**Comment by wheaton5** [comment]
Sometimes the download fails. wget from google drive is wonky and can fail silently. delete the current souporcell.sif and redownload. Thanks @drneavin hopefully this works

**Comment by SharonWang** [comment]
@wheaton5 @drneavin Thank you so much for your quick response.  I redownloaded the singularity again today and rerun the pipeline and still getting the same error msgs:

merging vcfs
running vartrix
running souporcell clustering
/opt/souporcell/souporcell/target/release/souporcell -k 6 -a SIGAA7_souporcell/alt.mtx -r SIGAA7_souporcell/ref.mtx --restarts 100 -b /home/xw251/rds/rds-bg200-hphi-gottgens/users/mq224/project_citeseq/100match_cellRangerOUTPUT/SIGAA7_output/outs/raw_feature_bc_matrix/barcodes.tsv --min_ref 4 --min_alt 4 --threads 24
running souporcell doublet detection
Traceback (most recent call last):
  File "/opt/souporcell/souporcell_pipeline.py", line 563, in <module>
    doublets(args, ref_mtx, alt_mtx, cluster_file)
  File "/opt/souporcell/souporcell_pipeline.py", line 509, in doublets
    subprocess.check_call(["troublet", "--alts", alt_mtx, "--refs", ref_mtx, "--clusters", cluster_file], stdout = dub, stderr = err)
  File "/usr/local/envs/py36/lib/python3.6/subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['troublet', '--alts', 'SIGAA7_souporcell/alt.mtx', '--refs', 'SIGAA7_souporcell/ref.mtx', '--clusters', 'SIGAA7_souporcell/clusters_tmp.tsv']' returned non-zero exit status 101.

and it seems that something goes wrong in the clustering step as I get clusters.err file as the following:

total loci used 30117
thread 9 iteration 0 done with -inf, best so far -inf
thread 22 iteration 0 done with -inf, best so far -inf
thread 19 iteration 0 done with -inf, best so far -inf
thread 3 iteration 0 done with -inf, best so far -inf
thread 21 iteration 0 done with -inf, best so far -inf
thread 16 iteration 0 done with -inf, best so far -inf
thread 7 iteration 0 done with -inf, best so far -inf
thread 1 iteration 0 done with -inf, best so far -inf
thread 11 iteration 0 done with -inf, best so far -inf
thread 17 iteration 0 done with -inf, best so far -inf
thread 13 iteration 0 done with -inf, best so far -inf
thread 6 iteration 0 done with -inf, best so far -inf
thread 8 iteration 0 done with -inf, best so far -inf
thread 15 iteration 0 done with -inf, best so far -inf
thread 12 iteration 0 done with -inf, best so far -inf
thread 14 iteration 0 done with -inf, best so far -inf
thread 10 iteration 0 done with -inf, best so far -inf
thread 4 iteration 0 done with -inf, best so far -inf
thread 23 iteration 0 done with -inf, best so far -inf
thread 18 iteration 0 done with -inf, best so far -inf
thread 5 iteration 0 done with -inf, best so far -inf
thread 2 iteration 0 done with -inf, best so far -inf
thread 20 iteration 0 done with -inf, best so far -inf
thread 0 iteration 0 done with -inf, best so far -inf
thread 22 iteration 1 done with -inf, best so far -inf
thread 19 iteration 1 done with -inf, best so far -inf
thread 9 iteration 1 done with -inf, best so far -inf
thread 21 iteration 1 done with -inf, best so far -inf
thread 16 iteration 1 done with -inf, best so far -inf
thread 3 iteration 1 done with -inf, best so far -inf
thread 1 iteration 1 done with -inf, best so far -inf
thread 11 iteration 1 done with -inf, best so far -inf
thread 12 iteration 1 done with -inf, best so far -inf
thread 17 iteration 1 done with -inf, best so far -inf
thread 10 iteration 1 done with -inf, best so far -inf
thread 4 iteration 1 done with -inf, best so far -inf
thread 15 iteration 1 done with -inf, best so far -inf
thread 7 iteration 1 done with -inf, best so far -inf
thread 8 iteration 1 done with -inf, best so far -inf
thread 18 iteration 1 done with -inf, best so far -inf
thread 14 iteration 1 done with -inf, best so far -inf
thread 13 iteration 1 done with -inf, best so far -inf
thread 6 iteration 1 done with -inf, best so far -inf
thread 5 iteration 1 done with -inf, best so far -inf
thread 0 iteration 1 done with -inf, best so far -inf
thread 20 iteration 1 done with -inf, best so far -inf
thread 23 iteration 1 done with -inf, best so far -inf
thread 2 iteration 1 done with -inf, best so far -inf
thread 9 iteration 2 done with -inf, best so far -inf
thread 19 iteration 2 done with -inf, best so far -inf
thread 22 iteration 2 done with -inf, best so far -inf
thread 21 iteration 2 done with -inf, best so far -inf
thread 12 iteration 2 done with -inf, best so far -inf
thread 1 iteration 2 done with -inf, best so far -inf
thread 11 iteration 2 done with -inf, best so far -inf
thread 16 iteration 2 done with -inf, best so far -inf
thread 15 iteration 2 done with -inf, best so far -inf
thread 3 iteration 2 done with -inf, best so far -inf
thread 10 iteration 2 done with -inf, best so far -inf
thread 4 iteration 2 done with -inf, best so far -inf
thread 17 iteration 2 done with -inf, best so far -inf
thread 18 iteration 2 done with -inf, best so far -inf
thread 14 iteration 2 done with -inf, best so far -inf
thread 7 iteration 2 done with -inf, best so far -inf
thread 8 iteration 2 done with -inf, best so far -inf
thread 0 iteration 2 done with -inf, best so far -inf
thread 13 iteration 2 done with -inf, best so far -inf
thread 20 iteration 2 done with -inf, best so far -inf
thread 6 iteration 2 done with -inf, best so far -inf
thread 23 iteration 2 done with -inf, best so far -inf
thread 5 iteration 2 done with -inf, best so far -inf
thread 2 iteration 2 done with -inf, best so far -inf
thread 19 iteration 3 done with -inf, best so far -inf
thread 1 iteration 3 done with -inf, best so far -inf
thread 21 iteration 3 done with -inf, best so far -inf
thread 9 iteration 3 done with -inf, best so far -inf
thread 12 iteration 3 done with -inf, best so far -inf
thread 22 iteration 3 done with -inf, best so far -inf
thread 11 iteration 3 done with -inf, best so far -inf
thread 15 iteration 3 done with -inf, best so far -inf
thread 16 iteration 3 done with -inf, best so far -inf
thread 3 iteration 3 done with -inf, best so far -inf
thread 4 iteration 3 done with -inf, best so far -inf
thread 10 iteration 3 done with -inf, best so far -inf
thread 17 iteration 3 done with -inf, best so far -inf
thread 7 iteration 3 done with -inf, best so far -inf
thread 14 iteration 3 done with -inf, best so far -inf
thread 18 iteration 3 done with -inf, best so far -inf
thread 8 iteration 3 done with -inf, best so far -inf
thread 20 iteration 3 done with -inf, best so far -inf
thread 5 iteration 3 done with -inf, best so far -inf
thread 6 iteration 3 done with -inf, best so far -inf
thread 1 iteration 4 done with -inf, best so far -inf
thread 23 iteration 3 done with -inf, best so far -inf
thread 0 iteration 3 done with -inf, best so far -inf
thread 2 iteration 3 done with -inf, best so far -inf
thread 13 iteration 3 done with -inf, best so far -inf
thread 19 iteration 4 done with -inf, best so far -inf
thread 22 iteration 4 done with -inf, best so far -inf
thread 9 iteration 4 done with -inf, best so far -inf
thread 12 iteration 4 done with -inf, best so far -inf
thread 21 iteration 4 done with -inf, best so far -inf
thread 11 iteration 4 done with -inf, best so far -inf
thread 3 iteration 4 done with -inf, best so far -inf
thread 16 iteration 4 done with -inf, best so far -inf
thread 15 iteration 4 done with -inf, best so far -inf
thread 4 iteration 4 done with -inf, best so far -inf
thread 7 iteration 4 done with -inf, best so far -inf
thread 14 iteration 4 done with -inf, best so far -inf
thread 18 iteration 4 done with -inf, best so far -inf
thread 17 iteration 4 done with -inf, best so far -inf
thread 10 iteration 4 done with -inf, best so far -inf
thread 6 iteration 4 done with -inf, best so far -inf
thread 5 iteration 4 done with -inf, best so far -inf
thread 20 iteration 4 done with -inf, best so far -inf
thread 0 iteration 4 done with -inf, best so far -inf
thread 8 iteration 4 done with -inf, best so far -inf
thread 2 iteration 4 done with -inf, best so far -inf
thread 23 iteration 4 done with -inf, best so far -inf
thread 13 iteration 4 done with -inf, best so far -inf
best total log probability = -inf

The log probability is -inf...

Would it be possible to know what happens?

Best regards,

Xiaonan Wang

**Comment by wheaton5** [comment]
Can you paste the first 3 lines of the alt.mtx and ref.mtx here?

**Comment by SharonWang** [comment]
@wheaton5 
Thank you so much for your quick response. Two of my samples ran successfully and 4 failed. The failed ones are the ones with k=6 as it was a pool of >= 6 samples.

alt.mtx
%%MatrixMarket matrix coordinate real general
% written by sprs
40634 6794880 4443593
1 447286 0
1 706499 0
1 1179350 0
1 1273258 0
1 1335709 0
1 1553175 0
1 1878383 0
1 1904606 0
1 2022146 0

ref.mtx
%%MatrixMarket matrix coordinate real general
% written by sprs
40634 6794880 4443593
1 447286 1
1 706499 1
1 1179350 1
1 1273258 1
1 1335709 1
1 1553175 1
1 1878383 1
1 1904606 1
1 2022146 1

Best regards,

Xiaonan Wang

**Comment by wheaton5** [comment]
After the first 3 lines, are there any non-0 entries in the 3rd column of the alt.mtx?

The log likelihood would be -inf if the likelihood is 0. I am trying to understand why this would be. Even if there is no data this shouldn't be the case. Have you tried running without known_genotypes?

**Comment by SharonWang** [comment]
@wheaton5 Thanks a lot for your quick response. Yes, there are 1700095 lines with 3rd column !=0. I will give it a try without known_genotypes.

**Comment by SharonWang** [comment]
@wheaton5 Sorry for the late reply and hope you had a nice weekend. I have shared the required files via google drive with you. There should be an invitation email sent to your google email address as whheaton@gmail.com. If you did not receive or if you would need more files and details, please do let me know.

Best regards,

Xiaonan Wang

**Comment by SharonWang** [comment]
@wheaton5 This was run without known genotypes, I put also uploaded the output vcf from souporcell in the same shared folder. 
souporcell_merged_sorted_vcf.vcf.gz: a3549c998e5f3f05538c808c3435522c
souporcell_merged_sorted_vcf.vcf.gz.tbi: 1d310d048f8a4aa0a8c5eae353855f3c

This is human PBMC sample, so I guess with known genotypes, the files from https://ftp.ncbi.nlm.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF/ should be used? Would it be possible to know the how much effect would it be providing different vcf files?

**Comment by SharonWang** [comment]
@wheaton5 I am wondering if souporcell would also test the best number of clusters like the figure 2d and 2g in your biorxiv paper. Would it be possible to know where I can find the information from the output to generate such plots? Thank you.

**Comment by wheaton5** [comment]
@SharonWang 
I ran the data you sent with the following command 
singularity exec souporcell.sif souporcell_pipeline.py -i bam.bam -b barcodes.tsv -f refdata-cellranger-GRCh38-3.0.0/fasta/genome.fa -t 12 -o sharon1 -k 6
and it worked just fine. What is your command exactly?

**Comment by SharonWang** [comment]
@wheaton5 Thanks a lot for checking. very appreciated. I have uploaded the command I used in the folder. It is the same as what you used. I am also running again to check if it is due to the cluster issue as last time I ran, 4 jobs failed and 2 successfully finished. I will let you know when the jobs are finished. Hopefully they run fine this time.

**Comment by SharonWang** [comment]
@wheaton5 The jobs have finished and ended up with the same errors. I am wondering if anything has been updated with the singularity and whether if it would be possible to have a version number added to the singularity file so that I would be able to see if it is the same one as what you used?

**Comment by wheaton5** [comment]
@SharonWang 
I update the version numbers on more major versions. I should get better about doing this. But we can check with md5sum

md5sum souporcell.sif
0f0c8cce01dedde183e1a1302c8b31b8  souporcell.sif

So if you get a different hex number then it's not the same version. And actually my test was not run through singularity (double checking by doing that now, waiting on results, but I made the singularity build after I updated master and it pulls from master so I doubt this will be the issue unless I screwed up somehow).

**Comment by wheaton5** [comment]
@SharonWang Yeah with the singularity build I got exactly the same results. No errors. Good clustering (in that each cell's best cluster is significantly better than it's second best). Not sure what is going on.

**Comment by SharonWang** [comment]
@wheaton5 The md5sum is the same. I tried it again, failed. So I tried with the souporcell conda env with the following command: 

python souporcell.py -a SIGAA7_souporcell/alt.mtx -r SIGAA7_souporcell/ref.mtx -b barcodes.tsv -k 6 -t 12 -o clusters_tmp.tsv

But I got the following error messages:
Traceback (most recent call last):
  File "/home/xw251/miniconda3/envs/souporcell/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1864, in _create_c_op
    c_op = c_api.TF_FinishOperation(op_desc)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Cannot reshape a tensor with 83495485440 elements to shape [1031012352,6] (6186074112 elements) for 'gradients/GatherV2_grad/Reshape' (op: 'Reshape') with input shapes: [6794880,2048,6], [2] and with input tensors computed as partial shapes: input[1] = [1031012352,6].

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/xw251/rds/rds-bg200-hphi-gottgens/users/xw251/programs/souporcell/souporcell.py", line 163, in <module>
    optimizer = tf.train.AdamOptimizer(learning_rate=0.1).minimize(cost)
  File "/home/xw251/miniconda3/envs/souporcell/lib/python3.6/site-packages/tensorflow/python/training/optimizer.py", line 403, in minimize
    grad_loss=grad_loss)
  File "/home/xw251/miniconda3/envs/souporcell/lib/python3.6/site-packages/tensorflow/python/training/optimizer.py", line 512, in compute_gradients
    colocate_gradients_with_ops=colocate_gradients_with_ops)
  File "/home/xw251/miniconda3/envs/souporcell/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py", line 158, in gradients
    unconnected_gradients)
  File "/home/xw251/miniconda3/envs/souporcell/lib/python3.6/site-packages/tensorflow/python/ops/gradients_util.py", line 731, in _GradientsHelper
    lambda: grad_fn(op, *out_grads))
  File "/home/xw251/miniconda3/envs/souporcell/lib/python3.6/site-packages/tensorflow/python/ops/gradients_util.py", line 403, in _MaybeCompile
    return grad_fn()  # Exit early
  File "/home/xw251/miniconda3/envs/souporcell/lib/python3.6/site-packages/tensorflow/python/ops/gradients_util.py", line 731, in <lambda>
    lambda: grad_fn(op, *out_grads))
  File "/home/xw251/miniconda3/envs/souporcell/lib/python3.6/site-packages/tensorflow/python/ops/array_grad.py", line 450, in _GatherV2Grad
    values = array_ops.reshape(grad, values_shape)
  File "/home/xw251/miniconda3/envs/souporcell/lib/python3.6/site-packages/tensorflow/python/ops/gen_array_ops.py", line 7715, in reshape
    "Reshape", tensor=tensor, shape=shape, name=name)
  File "/home/xw251/miniconda3/envs/souporcell/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "/home/xw251/miniconda3/envs/souporcell/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/home/xw251/miniconda3/envs/souporcell/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3616, in create_op
    op_def=op_def)
  File "/home/xw251/miniconda3/envs/souporcell/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 2027, in __init__
    control_input_ops)
  File "/home/xw251/miniconda3/envs/souporcell/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1867, in _create_c_op
    raise ValueError(str(e))
ValueError: Cannot reshape a tensor with 83495485440 elements to shape [1031012352,6] (6186074112 elements) for 'gradients/GatherV2_grad/Reshape' (op: 'Reshape') with input shapes: [6794880,2048,6], [2] and with input tensors computed as partial shapes: input[1] = [1031012352,6].

Sorry, I am completely new to tensorflow. But it would be very helpful if you could please have a look if anything can be done with it?

Thank you!

**Comment by wheaton5** [comment]
@SharonWang souporcell.py is the 1.0 version of clustering. The step by step portion of the README is updated to the 2.0 clustering method. The new clustering is a compiled binary in souporcell/target/release/souporcell. If you are running it from the github repo, you will need to compile it (instructions in README) first.

I really don't believe it is possible if you are doing things right for us to get different results using the same data, same reference, same singularity build. The code is completely deterministic. Any randomization I have seeded. Can you show me your singularity command again?

**Comment by SharonWang** [comment]
@wheaton5 singularity exec /home/xw251/rds/rds-bg200-hphi-gottgens/users/xw251/programs/souporcell.sif souporcell_pipeline.py -i /home/xw251/rds/rds-bg200-hphi-gottgens/users/xw251/Lab_Members/Xiaonan/Package_test/souporcell/test/possorted_genome_bam.bam -b /home/xw251/rds/rds-bg200-hphi-gottgens/users/xw251/Lab_Members/Xiaonan/Package_test/souporcell/test/barcodes.tsv -f /home/xw251/rds/rds-bg200-hphi-gottgens/users/xw251/Lab_Members/Xiaonan/Package_test/souporcell/test/genome.fa -t 24 -o SIGAA7_souporcell -k 6

**Comment by SharonWang** [comment]
@wheaton5 a quick update, I used the latest version of clustering this time, with the command line:
souporcell -a SIGAA7_souporcell/alt.mtx -r SIGAA7_souporcell/ref.mtx -b barcodes.tsv -k 6 -t 12 > clusters_tmp.tsv

It gave me the same results (-Inf of the log likelihood) as the singularity. (It is very strange indeed that with the same singularity, the results are different)

**Comment by wheaton5** [comment]
And what is the error you get with that? Also from which directory are you running this? Singularity only mounts directories downstream of the directory from which you are running it.

**Comment by SharonWang** [comment]
@wheaton5 
thread 11 iteration 0 done with -inf, best so far -inf
thread 5 iteration 0 done with -inf, best so far -inf
thread 9 iteration 0 done with -inf, best so far -inf
thread 3 iteration 0 done with -inf, best so far -inf
thread 6 iteration 0 done with -inf, best so far -inf
thread 7 iteration 0 done with -inf, best so far -inf
thread 1 iteration 0 done with -inf, best so far -inf
thread 4 iteration 0 done with -inf, best so far -inf
thread 10 iteration 0 done with -inf, best so far -inf
thread 8 iteration 0 done with -inf, best so far -inf
thread 2 iteration 0 done with -inf, best so far -inf
thread 0 iteration 0 done with -inf, best so far -inf
thread 3 iteration 1 done with -inf, best so far -inf
thread 11 iteration 1 done with -inf, best so far -inf
thread 6 iteration 1 done with -inf, best so far -inf
thread 7 iteration 1 done with -inf, best so far -inf
thread 5 iteration 1 done with -inf, best so far -inf
thread 1 iteration 1 done with -inf, best so far -inf
thread 10 iteration 1 done with -inf, best so far -inf
thread 9 iteration 1 done with -inf, best so far -inf
thread 4 iteration 1 done with -inf, best so far -inf
thread 8 iteration 1 done with -inf, best so far -inf
thread 2 iteration 1 done with -inf, best so far -inf
thread 0 iteration 1 done with -inf, best so far -inf
thread 7 iteration 2 done with -inf, best so far -inf
thread 1 iteration 2 done with -inf, best so far -inf
thread 6 iteration 2 done with -inf, best so far -inf
thread 3 iteration 2 done with -inf, best so far -inf
thread 5 iteration 2 done with -inf, best so far -inf
thread 9 iteration 2 done with -inf, best so far -inf
thread 4 iteration 2 done with -inf, best so far -inf
thread 10 iteration 2 done with -inf, best so far -inf
thread 11 iteration 2 done with -inf, best so far -inf
thread 8 iteration 2 done with -inf, best so far -inf
thread 2 iteration 2 done with -inf, best so far -inf
thread 0 iteration 2 done with -inf, best so far -inf
thread 7 iteration 3 done with -inf, best so far -inf
thread 9 iteration 3 done with -inf, best so far -inf
thread 1 iteration 3 done with -inf, best so far -inf
thread 3 iteration 3 done with -inf, best so far -inf
thread 10 iteration 3 done with -inf, best so far -inf
thread 11 iteration 3 done with -inf, best so far -inf
thread 6 iteration 3 done with -inf, best so far -inf
thread 4 iteration 3 done with -inf, best so far -inf
thread 5 iteration 3 done with -inf, best so far -inf
thread 8 iteration 3 done with -inf, best so far -inf
thread 0 iteration 3 done with -inf, best so far -inf
thread 2 iteration 3 done with -inf, best so far -inf
thread 7 iteration 4 done with -inf, best so far -inf
thread 1 iteration 4 done with -inf, best so far -inf
thread 3 iteration 4 done with -inf, best so far -inf
thread 11 iteration 4 done with -inf, best so far -inf
thread 9 iteration 4 done with -inf, best so far -inf
thread 6 iteration 4 done with -inf, best so far -inf
thread 0 iteration 4 done with -inf, best so far -inf
thread 10 iteration 4 done with -inf, best so far -inf
thread 4 iteration 4 done with -inf, best so far -inf
thread 2 iteration 4 done with -inf, best so far -inf
thread 5 iteration 4 done with -inf, best so far -inf
thread 8 iteration 4 done with -inf, best so far -inf
thread 1 iteration 5 done with -inf, best so far -inf
thread 3 iteration 5 done with -inf, best so far -inf
thread 9 iteration 5 done with -inf, best so far -inf
thread 7 iteration 5 done with -inf, best so far -inf

The error messages are the same as what I have got from running singularity. But this is not run with singularity, this was run under conda souporcell environment.

**Comment by SharonWang** [comment]
@wheaton5 I am wondering if it would be possible to have the top 20 lines of your alt.mtx and ref.mtx? I just would like to compare if the results are fine till the allele counting step. Thanks.

**Comment by wheaton5** [comment]
alt.mtx
%%MatrixMarket matrix coordinate real general
% written by sprs
56492 8831 5947425
1 359 0
1 518 0
1 786 0
1 888 1
1 940 0
1 1443 1
1 1541 1
1 1547 0
1 1558 0
1 1805 0
1 1846 1
1 2208 1
1 2405 0
1 2576 0
1 2762 0
1 2826 0
1 2853 2

ref.mtx
%%MatrixMarket matrix coordinate real general
% written by sprs
56492 8831 5947425
1 359 1
1 518 1
1 786 1
1 888 0
1 940 1
1 1443 0
1 1541 0
1 1547 1
1 1558 1
1 1805 1
1 1846 0
1 2208 0
1 2405 1
1 2576 1
1 2762 1
1 2826 1
1 2853 0

**Comment by SharonWang** [comment]
@wheaton5 Thank you for your quick response. They are different unfortunately. Must be something strange happened before this step then.. if there is no randomness involved in this step..

**Comment by wheaton5** [comment]
@SharonWang 
so before clustering the alt.mtx and ref.mtx are different. ok back one more step to freebayes
zcat sharon2/souporcell_merged_sorted_vcf.vcf.gz | wc -l
56749

wait a min. Looking back at the alt.mtx and ref.mtx you posted previously the first lines were
%%MatrixMarket matrix coordinate real general
% written by sprs
40634 6794880 4443593

so those numbers on the first non-header line are
loci cells total_entries

So somehow you are running with 6794880 cells in your barcodes.tsv file?

**Comment by wheaton5** [comment]
post the first lines of your barcodes.tsv again

**Comment by SharonWang** [comment]
@wheaton5 Sorry, I found the problem, I copied the raw barcode matrix into the folder. That's why it is not running. Really sorry about this! I will try again with the filtered ones. Thanks a lot for your help.

**Comment by wheaton5** [comment]
Cool. Well I should probably fix this anyway and exit with a reasonable error instead of giving -inf values.

**Comment by SharonWang** [comment]
@wheaton5 thank you so much for your help. The jobs have finished and the results look really nice!

**Comment by Zethson** [comment]
@wheaton5 I seem to be running into the same issue:

```
Traceback (most recent call last):
  File "/opt/souporcell/souporcell_pipeline.py", line 595, in <module>
    doublets(args, ref_mtx, alt_mtx, cluster_file)
  File "/opt/souporcell/souporcell_pipeline.py", line 540, in doublets
    subprocess.check_call([directory+"/troublet/target/release/troublet", "--alts", alt_mtx, "--refs", ref_mtx, "--clusters", cluster_file], stdout = dub, stderr = err)
  File "/usr/local/envs/py36/lib/python3.6/subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/opt/souporcell/troublet/target/release/troublet', '--alts', 'souporcelloutput/alt.mtx', '--refs', 'souporcelloutput/ref.mtx', '--clusters', 'souporcelloutput/clusters_tmp.tsv']' returned non-zero exit status 101.
```

md5sum

```
md5sum souporcell.sif                                                                                                                         (base) 
7d04b95d64c3a6f039da33b7acc05c8d  souporcell.sif
```

Run command:
`singularity exec -H /mnt/platte_1 souporcell.sif souporcell_pipeline.py -i merged.bam -b barcodes.tsv -f hg38.fasta -t 8 -o souporcelloutput -k 4`

I would strongly appreciate help and guidance. Thank you very much!

**Comment by wheaton5** [comment]
It's actually the previous step that failed and output a 0 length clusters file. Can you look at clusters.err?

**Comment by Zethson** [comment]
> It's actually the previous step that failed and output a 0 length clusters file. Can you look at clusters.err?

Thank you very much for your quick response. I hope that we can resolve this.

Actually, there is no clusters.err file. Only clusters.done with is empty.

**Comment by wheaton5** [comment]
wc -l alt.mtx
Huh says we have loci, but maybe no counts on those loci. take a look at ref.mtx and alt.mtx and make sure there are some counts in them?

**Comment by Zethson** [comment]
> lukas@agando-MS-7B09 /m/p/m/souporcelloutput> wc -l alt.mtx                                                                                                            (single_cell_analysis) 
> 15266432 alt.mtx

**Comment by wheaton5** [comment]
is there a clusters.tsv or a clusters_tmp.tsv or anything like that? can you just ls -lah the otuput directory for me?

**Comment by Zethson** [comment]
@wheaton5 sure. Anything that helps us solves this:

```
-rw-rw-r-- 1 lukas lukas 242M Feb 25 07:26 alt.mtx
-rw-rw-r-- 1 lukas lukas   77 Feb 25 06:27 bcftools.err
-rw-rw-r-- 1 lukas lukas    0 Feb 25 11:43 clustering.done
-rw-rw-r-- 1 lukas lukas 5.6K Feb 25 11:43 clusters.err
-rw-rw-r-- 1 lukas lukas    0 Feb 25 11:42 clusters_tmp.tsv
-rw-rw-r-- 1 lukas lukas    0 Feb 25 12:10 clusters.tsv
-rw-rw-r-- 1 lukas lukas  183 Feb 25 12:10 doublets.err
-rw-rw-r-- 1 lukas lukas 8.3K Feb 25 00:16 fastqs.done
-rw-rw-r-- 1 lukas lukas 8.1K Feb 25 03:14 minimap.err
-rw-rw-r-- 1 lukas lukas 242M Feb 25 07:26 ref.mtx
-rw-rw-r-- 1 lukas lukas  368 Feb 25 03:14 remapping.done
-rw-rw-r-- 1 lukas lukas  512 Feb 25 04:50 retag.err
-rw-rw-r-- 1 lukas lukas    0 Feb 25 06:05 retagging.done
-rw-rw-r-- 1 lukas lukas  39M Feb 25 06:27 souporcell_merged_sorted_vcf.vcf.gz
-rw-rw-r-- 1 lukas lukas 465K Feb 25 06:27 souporcell_merged_sorted_vcf.vcf.gz.tbi
-rw-rw-r-- 1 lukas lukas  20G Feb 25 05:59 souporcell_minimap_tagged_sorted.bam
-rw-rw-r-- 1 lukas lukas 4.6M Feb 25 06:05 souporcell_minimap_tagged_sorted.bam.bai
-rw-rw-r-- 1 lukas lukas   53 Feb 25 06:27 variants.done
-rw-rw-r-- 1 lukas lukas    0 Feb 25 07:26 vartrix.done
```

Both cluster files are empty.

**Comment by wheaton5** [comment]
top alt.mtx

**Comment by Zethson** [comment]
```
lukas@agando-MS-7B09 /m/p/m/souporcelloutput> head alt.mtx                                                                                                             (single_cell_analysis) 
%%MatrixMarket matrix coordinate real general
% written by sprs
409081 6794880 15266429
1 30565 0
1 36962 0
1 157722 0
1 243539 0
1 275510 0
1 325508 0
1 409776 0

```

**Comment by wheaton5** [comment]
right, so you are running with 15266429 cells in your barcodes.tsv file. Souporcell is meant to run with cell barcodes according to the output of cellranger. So you should have somewhere in the 1k-20k lines in your barcodes file.

**Comment by Zethson** [comment]
> right, so you are running with 15266429 cells in your barcodes.tsv file. Souporcell is meant to run with cell barcodes according to the output of cellranger. So you should have somewhere in the 1k-20k lines in your barcodes file.

Ahh, I see. So we lost the barcodes.tsv file of cell ranger and I "manually" recreated it. So yes, this may indeed be the issue. I will recreate the barcodes and report back.

Thank you very much for your help already!
